### ISSUE TITLE
Possible deadlock with multiprocessing

### ISSUE THREAD

**Issue by refraction-ray** [tensorflow/quantum#335]
See code demo below
```python
import tensorflow as tf
from multiprocessing import Pool
import cirq
import tensorflow_quantum as tfq
from sympy import symbols

def parallel_kernel(a):
    """tensorflow kernel"""
    a = a[0]
    m = tf.keras.Sequential([tf.keras.layers.Dense(8), tf.keras.layers.Dense(1)])
    with tf.GradientTape() as t:
        loss = m(a*tf.ones([32, 2]))
    g = t.gradient(loss, m.variables)
    return loss, g


def tfq_parallel_kernel(preset):
    """tensorflow quantum kernel"""
    ci = cirq.Circuit()
    ep = tfq.layers.Expectation()
    v = symbols("v_{0:9}")
    for i in range(2):
        ci.append(cirq.rx(v[i])(cirq.GridQubit(i, 0)))
    pnnp = tf.constant(preset)
    with tf.GradientTape() as t:
        t.watch(pnnp)
        loss = ep(
            inputs=[ci],
            symbol_names=v[: 2],
            symbol_values=[pnnp],
            operators=[cirq.X(cirq.GridQubit(0,0))],
        )[0]
        print(loss)
        gr = t.gradient(loss, pnnp)
    return loss, gr


def main():
    tf.keras.Sequential([tf.keras.layers.Input(1),tf.keras.layers.Dense(8), tf.keras.layers.Dense(1)])
    r = tfq_parallel_kernel([1, 1.2])
    print("Sequential:" , r)
    pool = Pool(2)
    r = pool.starmap(parallel_kernel, [([1., 1.2],), ([2., 1.8],)])
    print("tf Parallel mode", r)
    r = pool.starmap(tfq_parallel_kernel, [([1., 1.2],), ([2., 1.8],)])
    print("tfq Parallel mode", r)
    pool.close()
    return r

main()

```

The above code will stall in `tfq_parallel_kernel` in the `t.gradient` step,  with the exact stall position at
```
    main()
  File "tf_para_demo.py", line 43, in main
    r = pool.starmap(tfq_parallel_kernel, [([1., 1.2],), ([2., 1.8],)])
  File "/home/*/.conda/envs/tf2/lib/python3.6/multiprocessing/pool.py", line 274, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
  File "/home/*/.conda/envs/tf2/lib/python3.6/multiprocessing/pool.py", line 638, in get
    self.wait(timeout)
  File "/home/*/.conda/envs/tf2/lib/python3.6/multiprocessing/pool.py", line 635, in wait
    self._event.wait(timeout)
  File "/home/*/.conda/envs/tf2/lib/python3.6/threading.py", line 551, in wait
    signaled = self._cond.wait(timeout)
  File "/home/*/.conda/envs/tf2/lib/python3.6/threading.py", line 295, in wait
    waiter.acquire()
```
i.e. the program has trouble with lock in multiprocessing where possible deadlock happens.

Some further aspects:

1. This bug is specifically related to tfq instead of tf, as plain `parallel_kernel` with only tensorflow works well.

2. This bug is related to some keras model initialization in the main process (as the first line in ``main`` in the above code). Although such model has nothing to do with logic in multiprocessing or ``tfq_parallel_kernel``, the program stalls. If we delete the first line of ``main`` (initialization on a plain Keras sequential model), the program works fine.

Any thoughts on this?

**Comment by refraction-ray** [comment]
I have just read this: https://pythonspeed.com/articles/python-multiprocessing/ and https://github.com/pytorch/pytorch/issues/3492. I found the following workaround works, I will test further whether this works in my real project.

```python
import tensorflow as tf
from multiprocessing import set_start_method
from multiprocessing import get_context
import cirq
import tensorflow_quantum as tfq
from sympy import symbols

def parallel_kernel(a):
    a = a[0]
    m = tf.keras.Sequential([tf.keras.layers.Dense(8), tf.keras.layers.Dense(1)])
    with tf.GradientTape() as t:
        loss = m(a*tf.ones([32, 2]))
    g = t.gradient(loss, m.variables)
    return loss, g


def tfq_parallel_kernel(preset):
    ci = cirq.Circuit()
    ep = tfq.layers.Expectation()
    v = symbols("v_{0:9}")
    for i in range(2):
        ci.append(cirq.rx(v[i])(cirq.GridQubit(i, 0)))
    pnnp = tf.constant(preset)
    with tf.GradientTape() as t:
        t.watch(pnnp)
        loss = ep(
            inputs=[ci],
            symbol_names=v[: 2],
            symbol_values=[pnnp],
            operators=[cirq.X(cirq.GridQubit(0,0))],
        )[0]
        print(loss)
        gr = t.gradient(loss, pnnp)
    return loss, gr


def main():
    tf.keras.Sequential([tf.keras.layers.Input(2), tf.keras.layers.Dense(8), tf.keras.layers.Dense(1)])
    r = tfq_parallel_kernel([1, 1.2])
    print("Sequential:" , r)
    with get_context("spawn").Pool() as pool:
        r = pool.starmap(parallel_kernel, [([1., 1.2],), ([2., 1.8],)])
        print("tf Parallel mode", r)
        r = pool.starmap(tfq_parallel_kernel, [([1., 1.2],), ([2., 1.8],)])
    print("tfq Parallel mode", r)
    return r

if __name__ == "__main__":
    set_start_method("spawn")
    main()
```

The changes are:
1. We use `` set_start_method("spawn")`` and the pool is invoked by ``with get_context("spawn").Pool() as pool`` instead of plain Pool backend by ``fork``.
2. Note there must be ``if __name__ == "__main__":`` and ``set_start_method`` must be within the main block.

**Comment by Zhouquan-Wan** [comment]
According to https://github.com/pytorch/pytorch/issues/3492 you may also add `force=True` as `set_start_method('spawn', force=True)`.

**Comment by mhucka** [comment]
In the interest of prioritizing work on TensorFlow Quantum, I will close this issue as something that's been addressed. Please reopen this issue if that's incorrect.
